src/audio/chain_dma.c

```diff
@@ -0,0 +1,596 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Piotr Makaruk <piotr.makaruk@intel.com>
+
+#include <sof/audio/chain_dma.h>
+#include <ipc4/error_status.h>
+#include <ipc4/module.h>
+#include <sof/ut.h>
+#include <sof/lib/dma.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/drivers/dma.h>
+
+static const struct comp_driver comp_chain_dma;
+
+LOG_MODULE_REGISTER(chain_dma, CONFIG_SOF_LOG_LEVEL);
+
+/* 6a0a274f-27cc-4afb-a3e7-3444723f432e */
+DECLARE_SOF_RT_UUID("chain_dma", chain_dma_uuid, 0x6a0a274f, 0x27cc, 0x4afb,
+		    0xa3, 0xe7, 0x34, 0x44, 0x72, 0x3f, 0x43, 0x2e);
+DECLARE_TR_CTX(chain_dma_tr, SOF_UUID(chain_dma_uuid), LOG_LEVEL_INFO);
+
+static struct comp_dev *chain_task_create(const struct comp_driver *drv,
+					 const struct comp_ipc_config *ipc_config,
+					 const void *ipc_specific_config)
+{
+	const struct ipc4_chain_dma *cdma = (struct ipc4_chain_dma *)ipc_specific_config;
+	const uint32_t host_dma_id = cdma->primary.r.host_dma_id;
+	const uint32_t link_dma_id = cdma->primary.r.link_dma_id;
+	const uint32_t fifo_size = cdma->extension.r.fifo_size;
+	const bool scs = cdma->primary.r.scs;
+	struct chain_dma_data *cd;
+	struct comp_dev *dev;
+	int ret;
+
+	if (host_dma_id >= max_chain_number)
+		return NULL;
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		goto error1;
+
+	cd->first_data_received = false;
+	cd->under_over_run_notification_sent = false;
+	cd->cs = scs ? 2 : 4;
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+
+	comp_set_drvdata(dev, cd);
+
+	ret = chain_task_init(dev, host_dma_id, link_dma_id, fifo_size);
+	if (ret)
+		goto error2;
+
+	return dev;
+
+error2:
+	rfree(cd);
+	goto error1;
+error1:
+	rfree(dev);
+	return NULL;
+}
+
+int chain_task_start(struct comp_dev *dev)
+{
+	struct comp_driver_list *drivers = comp_drivers_get();
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	k_spinlock_key_t key;
+	int ret;
+
+	comp_info(dev, "chain_task_start(), host_dma_id = 0x%08x", cd->host_connector_node_id.dw);
+
+	key = k_spin_lock(&drivers->lock);
+	switch (cd->chain_task.state) {
+	case SOF_TASK_STATE_QUEUED:
+		k_spin_unlock(&drivers->lock, key);
+		return 0;
+	case SOF_TASK_STATE_COMPLETED:
+		break;
+	case SOF_TASK_STATE_INIT:
+		break;
+	default:
+		comp_err(dev, "chain_task_start(), bad state transition");
+		k_spin_unlock(&drivers->lock, key);
+		return -EINVAL;
+	}
+	cd->under_over_run_notification_sent = false;
+
+	if (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK) {
+		ret = chain_host_start(dev);
+		if (ret)
+			goto error1;
+		ret = chain_link_start(dev);
+		if (ret)
+			goto error2;
+	} else
+	{
+		ret = chain_link_start(dev);
+		if (ret)
+			goto error1;
+		ret = chain_host_start(dev);
+		if (ret)
+			goto error3;
```

Comment:
error 3 is used only here. Maybe instead of jumping between labels call chain_link_stop here and jump to the error handler?